src/main/kotlin/se/uulm/snowballr/backend/service/CriterionService.kt

```diff
@@ -24,13 +48,137 @@ interface ICriterionService {
+        val bothNotNull = criterion.projectId != null && criterion.createdBy != null
+        if (bothNull || bothNotNull) {
+            throw SnowballRException.FailedPreconditionException("The criterion is not valid anymore.")
+        }
+    }
+
+    /**
+     * Checks if the given user has the required permissions to access or modify a project criterion.
+     *
+     * This function verifies if the user has permission to access or update a criterion
+     * that belongs to an active project and throws exceptions when access is unauthorized
+     * or when attempting to update a criterion of an inactive project.
+     *
+     * @param criterion The criterion for which permissions need to be checked. It may or may not
+     *                  be associated with a project.
+     * @param currentUser The user whose permissions are being validated.
+     * @param requestId The identifier of the request being performed, used in exception messages.
+     * @param accessType The type of access being requested (e.g., READ, UPDATE).
+     *
+     * @throws SnowballRException.UnauthorizedException If the user does not have access permissions
+     *         to the criterion.
+     * @throws SnowballRException.FailedPreconditionException If attempting to update a criterion
+     *         in a project that is not active.
+     */
+    private suspend fun checkProjectCriterionPermission(
+        criterion: Criterion,
+        currentUser: User,
+        requestId: String,
+        accessType: AccessType,
+    ) {
+        if (criterion.projectId != null && criterion.createdBy == null) {
+            val project = projectRepo.getProjectById(criterion.projectId)
+            val members = when (accessType) {
+                AccessType.READ -> projectMemberRepo.getMembersOfProject(project.id)
+                AccessType.UPDATE -> projectMemberRepo.getAllProjectAdmins(project.id)
+                else -> emptyList()
+            }
+
+            if (!members.map { member -> member.userId }.contains(currentUser.id)) {
+                verifyServerAdminRole(currentUser) {
+                    throw UnauthorizedException.Single(
+                        EntityType.CRITERION,
+                        requestId,
+                        AccessType.READ,
+                        currentUser.id.toString(),
+                    )
+                }
+            }
+            if (accessType == AccessType.UPDATE &&
+                project.status != ProjectOuterClass.ProjectStatus.PROJECT_STATUS_ACTIVE
+            ) {
+                verifyServerAdminRole(currentUser) {
+                    throw SnowballRException.FailedPreconditionException(
+                        "The project is not active. Cannot update criterion with the ID: $requestId",
+                    )
+                }
+            }
+        }
+    }
+
+    /**
+     * Checks if the current user has permission to access or interact with the provided criterion.
```

Comment:
According to our `AccessType`, `access` is the general term for all CRUD operations.
```suggestion
     * Checks if the current user has permission to access the provided criterion.
```